rust/gitxetcore/src/upgrade_checks.rs

```diff
@@ -0,0 +1,431 @@
+        == Cmp::Gt
+}
+
+fn get_critical_text() -> String {
+    match std::env::var("_XET_TEST_VERSION_CHECK_CRITICAL_PATTERN") {
+        Ok(v) => v,
+        Err(_) => "CRITICAL".to_owned(),
+    }
+}
+
+fn get_version_info_filename() -> PathBuf {
+    let version_check_filename = match std::env::var("XET_UPGRADE_CHECK_FILENAME") {
+        Ok(v) => v,
+        Err(_) => VERSION_CHECK_FILENAME_HOME.to_owned(),
+    };
+
+    if let Some(mut path) = dirs::home_dir() {
+        path.push(version_check_filename);
+        return path;
+    }
+
+    if let Some(mut path) = dirs::cache_dir() {
+        path.push(version_check_filename);
+        return path;
+    }
+
+    // Guess it's just the local directory?
+    PathBuf::from(version_check_filename)
+}
+
+/// Queries the github releases API to get the latest release of the xet client.
+pub async fn get_newer_client_release_versions(
+    local_version: &str,
+    remote_repo_name: &str,
+) -> Option<Vec<(String, String)>> {
+    // Build the URL
+    let url =
+        format!("https://api.github.com/repos/{GITHUB_REPO_OWNER}/{remote_repo_name}/releases");
+
+    let mut headers = reqwest::header::HeaderMap::new();
+    headers.insert(
+        reqwest::header::USER_AGENT,
+        "XetData: xet-tools".parse().unwrap(),
+    );
+    headers.insert(
+        reqwest::header::ACCEPT,
+        "application/vnd.github.v3+json".parse().unwrap(),
+    );
+
+    let client = Client::new();
+    // Send the GET request
+    let Ok(response) = client.get(&url).headers(headers).send().await.map_err(|e|
+        {
+            info!("get_latest_client_release_version: Version check query returned error: {e:?}");
+            e
+        }) else { return None; };
+
+    // Ensure the request was successful
+    if let Ok(response) = response.error_for_status().map_err(|e| {
+        info!("get_latest_client_release_version: Server returned error: {e:?}");
+        e
+    }) {
+        let Ok(releases) = response.json::<Vec<Release>>().await.map_err(|e| {
+
+            info!("Error parsing info for response query as json: {e:?}");
+            e
+        }) else {return None; } ;
+
+        let mut newer_releases = Vec::new();
+
+        info!(
+            "get_newer_client_release_versions: Successfully retrieved {:?} new releases.",
+            releases.len()
+        );
+
+        // Check for all the releases that compare newer to this version string.
+        for release in releases {
+            let version_tag = release.tag_name.clone();
+
+            if version_is_newer(&version_tag, local_version) {
+                info!("get_newer_client_release_versions: Release {version_tag} newer than {local_version}.");
+                debug!(
+                    "get_newer_client_release_versions: Version {version_tag}; Body = {:?}",
+                    &release.body
+                );
+
+                newer_releases.push((version_tag, release.body));
+            } else {
+                break;
```

Comment:
Is it a valid assumption that returned releases are sorted in descending tag name?

Reply:
Yes, it is.